File: java/exercise1/SinglyLinkedList.java
package exercise1;

public class SinglyLinkedList<E> implements Cloneable {
  private static class Node<E> {
    private E element;
    private Node<E> next;

    public Node(E e, Node<E> n) {
      element = e;
      next = n;
    }

    public E getElement() {
      return element;
    }

    public Node<E> getNext() {
      return next;
    }

    public void setNext(Node<E> n) {
      next = n;
    }
  }

  private Node<E> head = null;
  private Node<E> tail = null;
  private int size = 0;

  public SinglyLinkedList() {
  }

  public int size() {
    return size;
  }

  public boolean isEmpty() {
    return size == 0;
  }

  public E first() {
    if (isEmpty())
      return null;
    return head.getElement();
  }

  public E last() {
    if (isEmpty())
      return null;
    return tail.getElement();
  }

  public void addFirst(E e) {
    head = new Node<>(e, head);
    if (size == 0)
      tail = head;
    size++;
  }

  public void addLast(E e) {
    Node<E> newest = new Node<>(e, null);
    if (isEmpty())
      head = newest;
    else
      tail.setNext(newest);
    tail = newest;
    size++;
  }

  public E removeFirst() {
    if (isEmpty())
      return null;
    E answer = head.getElement();
    head = head.getNext();
    size--;
    if (size == 0)
      tail = null;
    return answer;
  }

  @SuppressWarnings({ "unchecked" })
  public boolean equals(Object o) {
    if (o == null)
      return false;
    if (getClass() != o.getClass())
      return false;
    SinglyLinkedList other = (SinglyLinkedList) o;
    if (size != other.size)
      return false;
    Node walkA = head;
    Node walkB = other.head;
    while (walkA != null) {
      if (!walkA.getElement().equals(walkB.getElement()))
        return false;
      walkA = walkA.getNext();
      walkB = walkB.getNext();
    }
    return true;
  }

  @SuppressWarnings({ "unchecked" })
  public SinglyLinkedList<E> clone() throws CloneNotSupportedException {
    SinglyLinkedList<E> other = (SinglyLinkedList<E>) super.clone();
    if (size > 0) {
      other.head = new Node<>(head.getElement(), null);
      Node<E> walk = head.getNext();
      Node<E> otherTail = other.head;
      while (walk != null) {
        Node<E> newest = new Node<>(walk.getElement(), null);
        otherTail.setNext(newest);
        otherTail = newest;
        walk = walk.getNext();
      }
    }
    return other;
  }

  public int hashCode() {
    int h = 0;
    for (Node walk = head; walk != null; walk = walk.getNext()) {
      h ^= walk.getElement().hashCode();
      h = (h << 5) | (h >>> 27);
    }
    return h;
  }

  public String toString() {
    StringBuilder sb = new StringBuilder("(");
    Node<E> walk = head;
    while (walk != null) {
      sb.append(walk.getElement());
      if (walk != tail)
        sb.append(", ");
      walk = walk.getNext();
    }
    sb.append(")");
    return sb.toString();
  }

  public void swapTwoNodes(Node<E> node1, Node<E> node2) {
    if (node1 == node2) {
      return;
    }

    Node<E> prev1 = null;
    Node<E> prev2 = null;
    Node<E> curr1 = head;
    Node<E> curr2 = head;

    while (curr1 != null && curr1 != node1) {
      prev1 = curr1;
      curr1 = curr1.getNext();
    }

    while (curr2 != null && curr2 != node2) {
      prev2 = curr2;
      curr2 = curr2.getNext();
    }

    if (curr1 == null || curr2 == null) {
      return;
    }

    if (prev1 != null) {
      prev1.setNext(curr2);
    } else {
      head = curr2;
    }

    if (prev2 != null) {
      prev2.setNext(curr1);
    } else {
      head = curr1;
    }

    Node<E> temp = curr1.getNext();
    curr1.setNext(curr2.getNext());
    curr2.setNext(temp);

    if (curr1.getNext() == null) {
      tail = curr1;
    } else if (curr2.getNext() == null) {
      tail = curr2;
    }
  }

  public static void main(String[] args) {
    SinglyLinkedList<String> list = new SinglyLinkedList<String>();
    list.addFirst("MSP");
    list.addLast("ATL");
    list.addLast("BOS");
    list.addFirst("LAX");
    System.out.println(list);

    Node<String> node1 = list.head.getNext();
    Node<String> node2 = list.tail;
    list.swapTwoNodes(node1, node2);
    System.out.println(list);
  }
}

File: python/exercise1/SinglyLinkedlist.py
class Node:
    def __init__(self, element, next_node=None):
        self.element = element
        self.next_node = next_node


class SinglyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.size = 0

    def __len__(self):
        return self.size

    def is_empty(self):
        return self.size == 0

    def first(self):
        if self.is_empty():
            return None
        return self.head.element

    def last(self):
        if self.is_empty():
            return None
        return self.tail.element

    def add_first(self, e):
        newest = Node(e, next_node=self.head)
        self.head = newest
        if self.is_empty():
            self.tail = self.head
        self.size += 1

    def add_last(self, e):
        newest = Node(e)
        if self.is_empty():
            self.head = newest
        else:
            self.tail.next_node = newest
        self.tail = newest
        self.size += 1

    def remove_first(self):
        if self.is_empty():
            return None
        answer = self.head.element
        self.head = self.head.next_node
        self.size -= 1
        if self.is_empty():
            self.tail = None
        return answer

    def __eq__(self, other):
        if not isinstance(other, SinglyLinkedList) or self.size != len(other):
            return False

        node1, node2 = self.head, other.head
        while node1 is not None:
            if node1.element != node2.element:
                return False
            node1, node2 = node1.next_node, node2.next_node

        return True

    def __str__(self):
        result = []
        node = self.head
        while node is not None:
            result.append(str(node.element))
            node = node.next_node
        return "(" + ", ".join(result) + ")"

    def swapTwoNodes(self, node1, node2):
        if node1 == node2:
            return

        prevNode1 = prevNode2 = None
        currentNode = self.head

        while currentNode is not None:
            if currentNode.next_node == node1:
                prevNode1 = currentNode
            elif currentNode.next_node == node2:
                prevNode2 = currentNode
            currentNode = currentNode.next_node

        if prevNode1 is None and prevNode2 is None:
            return

        if prevNode1 is not None:
            prevNode1.next_node = node2
        else:
            self.head = node2

        if prevNode2 is not None:
            prevNode2.next_node = node1
        else:
            self.head = node1

        temp = node1.next_node
        node1.next_node = node2.next_node
        node2.next_node = temp

        if self.tail == node1:
            self.tail = node2
        elif self.tail == node2:
            self.tail = node1


if __name__ == "__main__":
    list = SinglyLinkedList()
    list.add_first("MSP")
    list.add_last("ATL")
    list.add_last("BOS")
    list.add_first("LAX")
    print(f"Before swapping: {list}")

    node1 = list.head.next_node
    node2 = list.tail
    list.swapTwoNodes(node1, node2)
    print(f"After swapping: {list}")
